Fix Cohere embedding error when base64 encoding_format is requested

colesmcintosh opened the pull request:
## Title

Fix Cohere embedding error when base64 encoding_format is requested

## Relevant issues

Fixes compatibility issues when using LangChain's OpenAI client with Cohere embeddings through LiteLLM.

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🐛 Bug Fix

## Changes

### Problem
When using LangChain's OpenAI embeddings client with Cohere models through LiteLLM, the client sends `encoding_format: "base64"` which is valid for OpenAI but not supported by Cohere. This causes Cohere's API to return an error: `"base64 embedding_type parameter is not supported"`.

### Solution
Modified the `map_openai_params` method in both `transformation.py` and `v1_transformation.py` to:
- Check if the requested `encoding_format` is a valid Cohere embedding type
- Only set the `embedding_types` parameter when valid formats are provided
- Silently ignore unsupported formats like "base64" instead of sending them to Cohere

### Testing
- Added unit test `test_cohere_embed_base64_encoding_format` to ensure base64 encoding format is handled gracefully
- Verified the fix works with the reproduction script
- Test passes locally:
```
poetry run pytest tests/llm_translation/test_cohere.py::test_cohere_embed_base64_encoding_format -xvs
```

<img width="1165" height="53" alt="Screenshot 2025-07-17 at 10 51 41 AM" src="https://github.com/user-attachments/assets/4b3de4ae-de1c-4856-bc79-b4ac7f40fa55" />

### Files Changed
- `litellm/llms/cohere/embed/transformation.py` - Updated map_openai_params to filter unsupported encoding formats
- `litellm/llms/cohere/embed/v1_transformation.py` - Same fix for v1 API compatibility
- `tests/llm_translation/test_cohere.py` - Added test case for base64 encoding format handling

colesmcintosh committed:
Fix Cohere embedding error when base64 encoding_format is requested

- Modified map_openai_params in both transformation.py and v1_transformation.py
- Only set embedding_types parameter when valid Cohere formats are specified
- Filter out unsupported formats like "base64" instead of sending them to Cohere API
- Added test to ensure base64 encoding_format is handled gracefully
- This fixes compatibility issues when using LangChain's OpenAI client with Cohere embeddings through LiteLLM

colesmcintosh committed:
Refactor Cohere embedding parameter handling in transformation.py and v1_transformation.py

- Simplified the filtering of valid Cohere encoding formats in the map_openai_params method.
- Removed redundant comments and improved clarity by directly filtering unsupported formats.
- Ensured only valid formats are passed to the optional_params for embedding_types, enhancing compatibility with Cohere API.

colesmcintosh committed:
Implement error handling for unsupported base64 encoding format in Cohere embeddings

- Added checks in map_openai_params to raise a CohereError when base64 encoding format is specified without enabling drop_params.
- Updated tests to verify that the error is raised correctly and that unsupported parameters are filtered out when drop_params=True is set.
- This enhances the robustness of the Cohere embedding configuration by providing clear feedback on unsupported formats.

krrishdholakia commented on litellm/llms/cohere/embed/transformation.py:
```diff
@@ -50,10 +50,23 @@ def map_openai_params(
+                # Check if base64 is in the formats
+                has_base64 = "base64" in formats
+                
+                if has_base64 and not drop_params:
```
support `litellm.drop_params` here as well

krrishdholakia commented on litellm/llms/cohere/embed/v1_transformation.py:
```diff
@@ -28,11 +30,26 @@ def get_supported_openai_params(self) -> List[str]:
+                # Check if base64 is in the formats
+                has_base64 = "base64" in formats
+                
+                if has_base64 and not drop_params:
```
support litellm.drop_params here

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/cohere-embedding-base64-parameter

colesmcintosh committed:
feat: Add global litellm.drop_params support for Cohere embeddings

Based on PR review feedback, enhanced the Cohere embedding parameter handling to:
- Check global litellm.drop_params setting in addition to the function parameter
- Add helpful documentation in error messages pointing to https://docs.litellm.ai/docs/completion/drop_params
- Add test case for global litellm.drop_params=True behavior

This ensures consistent behavior across all LiteLLM providers when handling unsupported parameters.

colesmcintosh commented on litellm/llms/cohere/embed/v1_transformation.py:
```diff
@@ -28,11 +30,26 @@ def get_supported_openai_params(self) -> List[str]:
         return ["encoding_format"]
 
     def map_openai_params(
```
Yes, the function signature change works as expected.

1. The `v1_transformation.py` file is used in a legacy flow (via `handler.py`) where `map_openai_params` is *not* called. Instead, the handler directly calls `_transform_request()`.

2. The only place where `map_openai_params` is used is in `utils.py`, and it already passes the `drop_params` argument (line 2649):

```python
optional_params = provider_config.map_openai_params(
    non_default_params=non_default_params,
    optional_params={},
    model=model,
    drop_params=litellm.drop_params is True or (drop_params is not None and drop_params is True),
)
```

3. The signature change is backward compatible:

   * The new parameter has a default (`drop_params: bool = False`)
   * Any existing calls without it will continue to work as before

4. Both `transformation.py` and `v1_transformation.py` now have consistent signatures, improving maintainability.

All three test cases are passing (error case, `drop_params=True`, and global `litellm.drop_params=True`), confirming the change works correctly.

krrishdholakia replied:
if this function isn't used, then why is it being modified

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/cohere-embedding-base64-parameter

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/cohere-embedding-base64-parameter

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/cohere-embedding-base64-parameter

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/cohere-embedding-base64-parameter

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/cohere-embedding-base64-parameter

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/cohere-embedding-base64-parameter
